File: 257. Binary Tree Paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        res = []
        self.dfs(root, '', res)
        return res
        
    def dfs(self, root, path, res):
        if root:
            path += str(root.val)
            if not root.left and not root.right:
                res.append(path)
            else:
                path += '->'
                self.dfs(root.left, path, res)
                self.dfs(root.right, path, res)

File: 257. Binary Tree Paths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<String> ans;
    public List<String> binaryTreePaths(TreeNode root) {
        ans = new LinkedList<>();
        dfs(root, "");
        return ans;
    }
    private void dfs(TreeNode root, String s) {
        if (root==null) {
            return;
        }
        s += Integer.toString(root.val);
        if (root.left==null && root.right==null) {
            ans.add(s);
        } else {
            s += "->";
            dfs(root.left, s);
            dfs(root.right, s);
        }
        
    }
}